### ISSUE TITLE
UPS REST "A shipment cannot have a KGS/IN or LBS/CM or OZS/CM as its unit of measurements"

### ISSUE THREAD

**Issue by sanmic** [magento/magento2#38618]
### Preconditions and environment

Magento version 2.4.7

### Steps to reproduce

Set European Union as Origin of the Shipment in Sales->Delivery Methods->UPS

### Expected result

UPS rates is visible in checkout and cart.

### Actual result

No UPS rates is visible in checkout and cart.

### Additional information

Seems like UnitOfMeasurement is hardcoded to "IN".
If you have European Union (other origins may experience the same problem) as Origin of the Shipment this will cause an error in the UPS REST request:
"A shipment cannot have a KGS/IN or LBS/CM or OZS/CM as its unit of measurements"

1147         foreach ($rowRequest->getPackages() as $package) {
1148             $rateParams['RateRequest']['Shipment']['Package'][] = [
1149                 "PackagingType" => [
1150                     "Code" => "{$params['48_container']}",
1151                     "Description" => "Packaging"
1152                 ],
1153                 "Dimensions" => [
1154                     "UnitOfMeasurement" => [
1155                         "Code" => "IN",
1156                         "Description" => "Inches"
1157                     ],
1158                     "Length" => "5",
1159                     "Width" => "5",
1160                     "Height" => "5"
1161                 ],
1162                 "PackageWeight" => [
1163                     "UnitOfMeasurement" => [
1164                         "Code" => "{$rowRequest->getUnitMeasure()}"
1165                     ],
1166                     "Weight" => "{$this->_getCorrectWeight($package['weight'])}"
1167                 ]
1168             ];
1169         }

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [X] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by sanmic** [comment]
@magento give me 2.4-develop instance - upcoming 2.4.x release

**Comment by sanmic** [comment]
@magento I am working on this

**Comment by sanmic** [comment]
https://github.com/magento give me 2.4-develop instance - upcoming 2.4.x release

**Comment by sanmic** [comment]
@magento give me 2.4-develop instance - upcoming 2.4.x release

**Comment by sanmic** [comment]
To reproduce:
Sales -> Delivery methods -> UPS
Set Origin of the Shipment to **Shipments Originating in the European Union**
Set Weight Unit to **KGS**

No UPS rates is visible in frontend.

If Weight Unit is set to LBS rates is visible in fronted.
I can see by myself that LBS is hardcoded in the core files.

![REST-EU-Origin-KGS1](https://github.com/magento/magento2/assets/5544888/bef778ef-3f1d-4132-859a-0bb26864c3bb)
![REST-EU-Origin-KGS2](https://github.com/magento/magento2/assets/5544888/79242626-1d4a-458d-a846-cbb2d0bc72ca)
![REST-EU-Origin-LGS1](https://github.com/magento/magento2/assets/5544888/a702c362-d732-4321-83c2-0a175b5b2d06)
![REST-EU-Origin-LGS2](https://github.com/magento/magento2/assets/5544888/91d6dc77-8393-49d3-8a4a-a24494b4a2d3)
![REST-US-Origin1](https://github.com/magento/magento2/assets/5544888/40ebb595-d4b3-41ff-9f93-f7e24d2ac50e)
![REST-US-Origin2](https://github.com/magento/magento2/assets/5544888/e84f3792-a034-434e-bb3c-0527d3d0d81c)

**Comment by sanmic** [comment]
Area: UPS shipping

**Comment by sanmic** [comment]
How can I add a Add label to the ticket?

"Add Area: XXXXX label to the ticket, indicating the functional areas it may be related to."

________________________________
Från: magento-deployment-service[bot] ***@***.***>
Skickat: den 16 april 2024 10:00
Till: magento/magento2 ***@***.***>
Kopia: sanmic ***@***.***>; Mention ***@***.***>
Ämne: Re: [magento/magento2] UPS REST "A shipment cannot have a KGS/IN or LBS/CM or OZS/CM as its unit of measurements" (Issue #38618)


Hi @sanmic<https://github.com/sanmic>, here is your Magento Instance: https://b2437ad179063c5f7eb8a7bbdd57dc08.instances-prod.magento-community.engineering
Admin access: https://b2437ad179063c5f7eb8a7bbdd57dc08.instances-prod.magento-community.engineering/admin_5a0f
Login: d7361ba1
Password: f4687c1b2cca

**Comment by engcom-Delta** [comment]
Hi @sanmic ,

 Verified the issue and the issue is reproducible.

 Hence, Confirming the issue.

 Steps to reproduce

   1. Sales -> shipping methods -> UPS 
   2. Enable UPS method
   3. Set Origin of the Shipment to Shipments Originating in the European Union
   4. Set Weight Unit to KGS
   5. Go to frontend and verify the shipping methods available
            
 
  Please refer the attached screenshot

<img width="1437" alt="Screenshot 2024-04-17 at 6 19 45 PM" src="https://github.com/magento/magento2/assets/51681379/f3e5bc51-8b0c-4392-aa94-54e1bba374e7">

<img width="1434" alt="Screenshot 2024-04-17 at 6 21 56 PM" src="https://github.com/magento/magento2/assets/51681379/223a7015-22df-4dd8-8c7d-a5f1e914ab6c">

<img width="1437" alt="Screenshot 2024-04-17 at 6 43 04 PM" src="https://github.com/magento/magento2/assets/51681379/03da8326-323f-458c-aac9-3873d3caaf4e">

<img width="1438" alt="Screenshot 2024-04-17 at 6 49 49 PM" src="https://github.com/magento/magento2/assets/51681379/42d363f2-3b1f-461e-9019-c76e88e378f4">

<img width="1440" alt="Screenshot 2024-04-17 at 6 51 41 PM" src="https://github.com/magento/magento2/assets/51681379/2da02bc3-2bae-47c9-9e84-6fc171d31ac7">

**Comment by engcom-Delta** [comment]
> @engcom-Delta How can I add a Add label to the ticket? "Add Area: XXXXX label to the ticket, indicating the functional areas it may be related to."

Hi @sanmic , I have added the labels on the ticket. Let us know if you need any changes.

**Comment by glo05363** [comment]
@sanmic I can reproduce this issue, while analysing the same, i can see that 
1. If we setup config with **LBS** we get the results for rates quotes:
<img width="1025" alt="Screenshot 2024-04-19 at 1 50 09 AM" src="https://github.com/magento/magento2/assets/87003041/cca5bb61-d082-4c20-ae14-9840c05e1508">

2. If we config KGS to get the Rate data we do not get the results.
<img width="1025" alt="Screenshot 2024-04-19 at 1 50 25 AM" src="https://github.com/magento/magento2/assets/87003041/10dc5fcd-47bb-477e-af27-33604a8c7927">

I am checking on how the **weight/Dimension** units are affecting the results as it is known the measurement units sometimes are not valid for the given country: Here with the same payload i have tried with both **LBS/KGS & INCHES/CM**

<img width="1025" alt="Screenshot 2024-04-19 at 1 57 14 AM" src="https://github.com/magento/magento2/assets/87003041/db3d2233-c72f-40b6-b4ca-72130488e854">

The analysis is in progress and thank you for this finding , we shall see on how to incorporate the dynamic Measurement units for Rates Quote Data, as we might need to implement the country based UPS support for the same.

**Comment by engcom-Bravo** [comment]
Hi @sanmic,

Our internal team is working on the issue and we have following updates

Please refer the Comments:

https://github.com/magento/magento2/issues/38618#issuecomment-2065269636


https://github.com/magento/magento2/issues/38618#issuecomment-2066513710


Thanks.

**Comment by digitalrisedorset** [comment]
@magento I am working on this

**Comment by digitalrisedorset** [comment]
Do we have some UPS access details to use to replicate the issue on a local environment?

**Comment by sanmic** [comment]
@digitalrisedorset Have you been able to get UPS access details?

**Comment by digitalrisedorset** [comment]
@sanmic no I have not. I got as far as creating the UPS account but a lot of settings were required and by the time I had the setup created to handle the task, I was doomed to have setup the UPS service differently to the needed to replicate the issue we have in this ticket. So, I'd like instead to verify if I can use a test UPS environment that is ready to use to just perform the Magento test and fix after this

**Comment by sanmic** [comment]
> @sanmic no I have not. I got as far as creating the UPS account but a lot of settings were required and by the time I had the setup created to handle the task, I was doomed to have setup the UPS service differently to what is needed to replicate the issue we have in this ticket. So, I'd like instead to verify if I can use a test UPS environment that is ready to use to just perform the Magento test and fix after this

@digitalrisedorset I guess I can create a UPS test app / credentials so you can continue your work.
I'll need a Callback URL and an email or similar to send the details.

**Comment by digitalrisedorset** [comment]
Hi @sanmic , I have setup a demo site at https://demo.digitalrisedorset.co.uk/ and the email may  be herve@digitalrisedorset.co.uk. Many thanks

**Comment by sanmic** [comment]
@digitalrisedorset Check you email :)

**Comment by digitalrisedorset** [comment]
I have setup UPS, to start with, I'd like to verify I can get it to work as per the screenshots above, I have setup the credentials with European origin and yet Pounds rather than Kgs. I have also used an address from the states. I attach both screenshots. Am I missing something?
![Screenshot from 2024-04-26 12-22-49](https://github.com/magento/magento2/assets/166377039/84e127bc-cffb-4dce-802b-1fc800e6794d)
![Screenshot from 2024-04-26 12-22-06](https://github.com/magento/magento2/assets/166377039/3dc0b007-41b9-411c-8aae-70f80dd8a9e3)

**Comment by sanmic** [comment]
@digitalrisedorset I think you have to set "Origin of the Shipment" to, for example, "Shipments Originating in United States".
The weight unit parameter may depend on the Magentos origin settings, Sales->Shipping settings->Origin, but I'm not 100% sure.
Our Origin, for example, is Sweden which uses KGS. If origin is US which use LBS.
Do some testing based on that, I think you will figure it out.

**Comment by digitalrisedorset** [comment]
I have looked into UPS guides further. I have worked out that the setup can go wrong with the information: User ID, Password, Gateway URL and origin setting:
- `User ID`: I have assigned the `Client ID` value
- `Password`: I have assigned the `Client Secret` value
- `Gateway URL`: I left this to the system value: `https://wwwcie.ups.com/api/rating/`
- **Origin and set up the right address**: I have `US, California, 90034` (there is no city, street address defined at the moment)

Could we double check these information settings before we go any further? 

#https://amasty.com/knowledge-base/how-to-set-up-ups-in-magento-2.html
#https://www.ups.com/upsdeveloperkit

**Comment by sanmic** [comment]
@digitalrisedorset I can confirm those settings work for me as long as I have LBS as weight unit.
Changing the "Origin of the Shipment" to "Europe", weight to KGS and the "Shipping Origin" to our address, "77463, Dalarnas Län, Avesta, Sweden", will break it.

I've just spent a short time on this, but this is what I've found out so far.
UPS will return rates if:
* Magento origin is set to a US address (will probably work with Canada and a lot of other countries as well)
* UPS weight unit is set to: LBS
* UPS origin setting doesn't matter if above is set.

UPS will NOT return rates if:
* UPS weight unit is set to; KGS

UPS will NOT return rates if:
* Magento origin is set to "77463, Dalarnas Län, Avesta, Sweden"

I want to setup my store like this:
* Magento origin: "77463, Dalarnas Län, Avesta, Sweden"
* UPS weight unit is set to: KGS
* UPS origin: Europe

I can change line 1155 in /vendor/magento/module-ups/Model/Carrier.php -> "Code" => "CM", to make it work with above settings.

**Comment by digitalrisedorset** [comment]
I have a draft PR in place for this issue. The changes in the PR do resolve the problem but I'd like to check if we can also ensure that the integration can work with the other modes without being restricted to Inches before I finalise the PR

see: https://github.com/magento/magento2/pull/38673

**Comment by digitalrisedorset** [comment]
@sanmic in the same way that Inch is hard-coded, I can see the dimensions for the request is also hard-coded. This is just below the place that you had patched:
```
"Length" => "5",
                    "Width" => "5",
                    "Height" => "5"
```

I have changed the part you recommended us to change but I have not changed the dimensions. Would you be able to comment whether we should change it or not?

**Comment by sanmic** [comment]
@digitalrisedorset Yes, I've noticed that also.
I cannot find an option to set or use dimensions in the UPS settings section so I guess the hardcoded 5x5x5 is there for some kind of fallback 🤔

**Comment by digitalrisedorset** [comment]
@sanmic the PR is now in review, I have tested today all the use cases you have mentioned and all is working with the simple changes added.

**Comment by digitalrisedorset** [comment]
@magento run all tests

**Comment by digitalrisedorset** [comment]
@magento run all tests

**Comment by hostep** [comment]
It looks like this was fixed in Magento 2.4.7-p1 which was released today: https://github.com/magento/magento2/commit/d10435b11ada4e502dca7539f8fd31d059d3c482#diff-8baae2f7a7af0d5144a7a6d168ad5e92161d596179feef9daa24ebefbb44d86d

**Comment by engcom-Charlie** [comment]
The required changes has been published in https://github.com/magento/magento2/commit/493e01f54c8922c620fe9c9f2bfdd1aa536b9bb6 for 2.4 develop.
